sched/event/event_wait.c

```diff
@@ -0,0 +1,213 @@
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxevent_tickwait
+ *
+ * Description:
+ *   Wait for all of the specified events from the specified tick time.
+ *
+ *   This routine waits on event object event until any of the specified
+ *   events have been delivered to the event object, or the maximum wait time
+ *   timeout has expired. A thread may wait on up to 32 distinctly numbered
+ *   events that are expressed as bits in a single 32-bit word.
+ *
+ * Input Parameters:
+ *   event  - Address of the event object
+ *   events - Set of events to post to event
+ *          - Set events to 0 will indicate wait from any events
+ *   delay  - Ticks to wait from the start time until the event is
+ *            posted.  If ticks is zero, then this function is equivalent
+ *            to nxevent_trywait().
+ *
+ * Returned Value:
+ *   This is an internal OS interface and should not be used by applications.
+ *   Return of matching events upon success.
+ *   0 if matching events were not received within the specified time.
+ *
+ ****************************************************************************/
+
+static nxevent_mask_t nxevent_wait_internal(FAR nxevent_t *event,
+                                            nxevent_mask_t events,
+                                            uint32_t delay)
+{
+  FAR struct tcb_s *rtcb;
+  irqstate_t flags;
+  int ret;
+
+  DEBUGASSERT(event != NULL && up_interrupt_context() == false);
+
+  flags = enter_critical_section();
+
+  /* Remove pending counter */
+
+  nxsem_trywait(&event->sem);
+
+  if ((events == 0 && event->events) ||
+      (events != 0 && event->events == events))
+    {
+      events = event->events;
+    }
+  else
+    {
+      rtcb         = nxsched_self();
+      rtcb->events = events;
+
+      if (delay == UINT32_MAX)
+        {
+          ret = nxsem_wait_uninterruptible(&event->sem);
+        }
+      else if (delay == 0)
+        {
+          ret = nxsem_trywait(&event->sem);
+        }
+      else
+        {
+          ret = nxsem_tickwait_uninterruptible(&event->sem, delay);
+        }
+
+      if (ret == 0)
+        {
+          if (events != 0)
+            {
+              events &= event->events;
+            }
+          else
+            {
+              events = event->events;
+            }
+        }
+      else
+        {
+          events = 0;
+        }
+    }
+
+  event->events = 0;
+
+  leave_critical_section(flags);
+
+  return events;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxevent_wait
+ *
+ * Description:
+ *   Post one or more events to an event object.
```

Comment:
wait

Reply:
Post one or more events to an event object.

Reply:
Done